spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -119,6 +150,70 @@ public Subscriber subscribe(String subscription, MessageReceiver messageHandler)
 		return subscriber;
 	}
 
+	/**
+	 * Pulls messages on demand using the pull request in argument.
+	 *
+	 * <p>
+	 * After messages are received, this method sends acknowledgements for the received messages if
+	 * any was received.
+	 */
+	@Override
+	public List<PubsubMessage> pull(PullRequest pullRequest) {
+		try {
+			try (SubscriberStub subscriber = GrpcSubscriberStub.create(this.pullSettings)) {
+				PullResponse pullResponse =	subscriber.pullCallable().call(pullRequest);
+
+				// Ack received messages.
+				if (pullResponse.getReceivedMessagesCount() > 0) {
+					List<String> ackIds = pullResponse.getReceivedMessagesList().stream()
+							.map(ReceivedMessage::getAckId)
+							.collect(Collectors.toList());
+
+					AcknowledgeRequest acknowledgeRequest = AcknowledgeRequest.newBuilder()
+							.setSubscriptionWithSubscriptionName(
+									pullRequest.getSubscriptionAsSubscriptionName())
+							.addAllAckIds(ackIds)
+							.build();
+
+					subscriber.acknowledgeCallable().call(acknowledgeRequest);
+				}
+
+				return pullResponse.getReceivedMessagesList().stream()
+						.map(ReceivedMessage::getMessage)
+						.collect(Collectors.toList());
+			}
+		}
+		catch (Exception ioe) {
+			throw new PubSubException("Error pulling messages from subscription "
+					+ pullRequest.getSubscription() + ".", ioe);
+		}
+	}
+
+	@Override
+	public List<PubsubMessage> pull(String subscription, Integer maxMessages,
+			Boolean returnImmediately) {
+		PullRequest.Builder pullRequestBuilder =
+				PullRequest.newBuilder().setSubscriptionWithSubscriptionName(
+						SubscriptionName.of(this.projectIdProvider.getProjectId(), subscription));
```

Comment:
I think we have to check `projectIdProvider` for null before performing its methods.